src/app/qgisapp.cpp

```diff
@@ -8003,6 +8003,33 @@ void QgisApp::saveEdits( QgsMapLayer *layer, bool leaveEditable, bool triggerRep
   if ( vlayer == activeLayer() )
     mSaveRollbackInProgress = true;
 
+  // wait for end of rendering to avoid lock in case of
+  // splilte and gpkg fix #15498
+  // Wait only if the current layer is in the rendering pool
+  if ( mMapCanvas->mapSettings().layers().contains( vlayer->id() ) )
+  {
+    if ( vlayer->dataProvider()->storageType().count( "SQLite" ) ||
+         vlayer->dataProvider()->storageType().count( "GPKG" ) )
+    {
+      // notify wait the rendering end
+      if ( mMapCanvas->isDrawing() )
+      {
+        // notify
+        QgsMessageBarItem *item = new QgsMessageBarItem( tr( "Save edit paused" ),
+            tr( "Waiting for end of rendering" ),
+            QgsMessageBar::WARNING, 0 );
+        messageBar()->pushItem( item );
```

Comment:
Probably just use `messageBar()->pushWarning(...)` instead of creating a new object, since the widget doesn't have any custom child widgets.

Reply:
or `messageBar()->pushMessage(...)` if you want no duration.

Reply:
Nevermind. Sorry, should have reviewed all the code first. I see you reference `QgsMessageBarItem` object below.

Reply:
exactly, I need the item reference to remove it directly after save commit is done

Reply:
If I get this right, you are basically adding a sleep in the main thread (that will block the GUI, even if it's for a short amount of time), have you considered adding a singleshot timer that re-triggers the same slot in case the canvas is drawing?  This would have the same effect (delaying the commit) but without blocking the GUI.

For master and C++11, a lambda would probably be the best way to implement it (see snippet below), but there might be ways to do that in 2.x too.

```
QTimer::singleShot(100, this, [ = ] {
          saveEdits( layer, leaveEditable, triggerRepaint );
        });
```

Also, instead of hardcoding the storage type, I would suggest to add a capability flag to the data provider class that indicates if there support for file lock (or the lack of it).

Reply:
Another idea: what about connecting to `connect(mMapCanvas, &QgsMapCanvas::renderComplete` ?

Reply:
I though about this kind of solution, but the main problem to introduce async management is that it would change/brake to "save editing" feature execution flow. In my implementation I just introduce a kind of not blocking "sleep" instead of restructuring the set of method calls. 
Another reason is to be conservative in bug fixing mode to avoid to introduce modification in the sequence of calls that can introduce more side effects (or regressions). Side effect not checked by unit or integrations tests because lack of them in the 2.x branch.

Reply:
About adding a capability, I also though about this, but I was not conformtable to enter a kind a capability that depend make sense only in some rendering cases. A Capability should be an objective property of a provider, so how to call it? isLockedDuringAccess() isLockedConcurrentwriteReadAccess() ? Because wasn't clear how to define this capability I preferred to avoid to introduce a new capability for only a bug fix.
I'm open to find a general capability semanttic that can fit this provider case.

Reply:
A delay/sleep in the main thread is blocking by definition and it's not very robust, because you'll never know if the delay is long enough.

Reply:
to use the main event loop I can single shooting QgisApp::saveEdits slot instead of creating a new event loop.
The main idea that I want to maintain is to avoid to change the current code flow.